executor/src/witgen/machines/block_machine.rs

```diff
@@ -409,23 +458,27 @@ impl<T: FieldElement> BlockMachine<T> {
         next: bool,
         constraint: Constraint<T>,
         change_logger: &mut ChangeLogger<T>,
-    ) {
+    ) -> bool {
         let r = (self.row + next as DegreeType) % self.degree;
         match constraint {
             Constraint::Assignment(a) => {
                 change_logger.log(self, &name, &a, Some(r));
+                let r = r as usize;
                 let values = self.data.get_mut(&poly).unwrap();
-                if (r as usize) < values.len() {
+                if r < values.len() && values[r].is_none() {
                     // do not write to other rows for now
-                    values[r as usize] = Some(a);
+                    values[r] = Some(a);
+                    true
+                } else {
+                    false
                 }
             }
-            Constraint::RangeConstraint(bc) => {
-                self.range_constraints
-                    .entry(poly)
-                    .or_default()
-                    .insert(r, bc);
-            }
+            Constraint::RangeConstraint(bc) => self
+                .range_constraints
+                .entry(poly)
+                .or_default()
+                .insert(r, bc)
+                .is_none(),
```

Comment:
Can it be none?

Reply:
Yes, if there was no range constraint for the polynomial & row combination before.

Reply:
I think it can happen that we already know a range constraint but it is not as strict. So instead of just `insert` we should use `and_modify` and compute and store the conjunction of the constraints.